smartsim/settings/dispatch.py

```diff
@@ -0,0 +1,321 @@
+# BSD 2-Clause License
+#
+# Copyright (c) 2021-2024, Hewlett Packard Enterprise
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from __future__ import annotations
+
+import dataclasses
+import subprocess as sp
+import typing as t
+import uuid
+
+from typing_extensions import Self, TypeAlias, TypeVarTuple, Unpack
+
+from smartsim._core.utils import helpers
+from smartsim.error import errors
+from smartsim.types import LaunchedJobID
+
+if t.TYPE_CHECKING:
+    from smartsim.experiment import Experiment
+    from smartsim.settings.arguments import LaunchArguments
+
+_Ts = TypeVarTuple("_Ts")
+_T_contra = t.TypeVar("_T_contra", contravariant=True)
+_DispatchableT = t.TypeVar("_DispatchableT", bound="LaunchArguments")
+_LaunchableT = t.TypeVar("_LaunchableT")
+
+_EnvironMappingType: TypeAlias = t.Mapping[str, "str | None"]
+_FormatterType: TypeAlias = t.Callable[
+    [_DispatchableT, "ExecutableProtocol", _EnvironMappingType], _LaunchableT
+]
+_LaunchConfigType: TypeAlias = (
+    "_LauncherAdapter[ExecutableProtocol, _EnvironMappingType]"
+)
+_UnkownType: TypeAlias = t.NoReturn
+
+
+@t.final
+class Dispatcher:
+    """A class capable of deciding which launcher type should be used to launch
+    a given settings type.
+    """
```

Comment:
Could you add a param desc here

Reply:
for the `__init__` method you mean? Of course!